common/rpc/grpc.go

```diff
@@ -181,20 +185,51 @@ func NewServiceErrorInterceptor(
 		}
 		return resp, serviceerror.ToStatus(err).Err()
 	}
+}
+
+func NewFrontendErrorInterceptor(
+	logger log.Logger, hideErrorsAtFrontend bool,
+) grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		_ *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (interface{}, error) {
 
+		resp, err := handler(ctx, req)
+		if err != nil && hideErrorsAtFrontend {
+			err = HideUnknownOrInternalErrors(logger, err)
+		}
+		return resp, serviceerror.ToStatus(err).Err()
+	}
 }
 
-func FrontendErrorInterceptor(
-	ctx context.Context,
-	req interface{},
-	_ *grpc.UnaryServerInfo,
-	handler grpc.UnaryHandler,
-) (interface{}, error) {
-	resp, err := handler(ctx, req)
+var errorFrontendUnknown = errors.New("something went wrong, please retry")
 
-	// mask some internal errors at frontend
+func HideUnknownOrInternalErrors(
+	logger log.Logger,
+	err error,
+) error {
+	// convert some internal errors into specific errors
 	if _, ok := err.(*serviceerrors.ShardOwnershipLost); ok {
-		err = serviceerror.NewUnavailable("shard unavailable, please backoff and retry")
+		return serviceerror.NewUnavailable("shard unavailable, please backoff and retry")
 	}
-	return resp, serviceerror.ToStatus(err).Err()
+
+	s := serviceerror.ToStatus(err)
+	if s.Code() != codes.Unknown && s.Code() != codes.Internal {
+		return err
+	}
+
+	// convert internal and unknown errors into neutral error with hash code of the original error
+	errorHash := errorMessageHash(err)
+	// log original error with hash
+	message := fmt.Sprintf("%s (%s)", err.Error(), errorHash)
+	logger.Error(message)
```

Comment:
This is an interesting idea. We also has this interceptor https://github.com/temporalio/temporal/blob/ee3e4925cd4fd602df3ab2d8c742a1240f5dd002/common/rpc/interceptor/telemetry.go#L397-L411. To avoid double logging, maybe it is better to do the error message hide there? We need to log the original and the masked error message, so we will be able to investigate when user report error with that hashcode.

Reply:
I was not aware about this interceptor. I will make a change (why log masked error message?).